Network/Socket/Types.hsc

```diff
@@ -70,14 +73,47 @@ import Network.Socket.Imports
 -----------------------------------------------------------------------------
 
 -- | Basic type for a socket.
-newtype Socket = Socket CInt deriving (Eq, Show)
+newtype Socket = Socket (IORef CInt)
+
+instance Show Socket where
+    show s = "<socket " ++ show (unsafeFdSocket s) ++ ">"
```

Comment:
Not sure that `unsafeFdSocket s` is a good idea here.  Other than distinguishing different sockets, the actual value of `fd` is not actually that interesting.

To me, it would make more sense to show the Socket type and address family, and ideally if bound the local address, if connected remote address...  But some of those properties mutate over time and require IO to query.  Perhaps, to avoid the I/O barrier the "show" string could be computed once when the socket is created?  Sockets are expensive enough to create that carrying an extra string should not be burdensome.  The string could be a ByteString if that's more efficient to store.

```
   <socket STREAM AF_INET _fd_>
   <socket DGRAM AF_INET6 _fd_>
   <socket STREAM UNIX _fd_>
   ...
```

Are there backwards compatibility requirements to retain the legacy Show format?  Note, I'm not sure whether my alternative suggestion is wise. 

Perhaps a minimal `Show` implementation is fine, and a more complete socket description could be provided as an IO action, that would also display the local address, the remote address (if connected), ...

```
displaySocket :: Socket -> IO String
```

Reply:
Perhaps even a `ShortByteString`

Reply:
Either way, I think it is best to avoid `unsafePerformIO` in the `Show` implementation.  Precomputing the string also means that we can display it even for closed sockets (where `it` is the description of the socket when it was opened)

Reply:
Precomputing is done in 40801af4b9e49a88c52e7928ec4bdf6c50361e60
I ensured that show string is exactly the same as the previous one.